Stricter parsing of relative urls

MatMoore opened the pull request:
I noticed that when passing a base URL to awesome bot, it picks up things that aren't URLs.

For example, I put some java code in a code block, and awesome bot reported it as a bad URI:

```java
public class HelloWorld {
	public static void main(String args[]){
		System.out.println("Hello World");
	}
}
```

I've changed the regex so that the link target has to directly follow the link text, and added some more tests.

The link parsing logic is still quite complex, and there are edge cases that still won't work, so it may be worth replacing it with a markdown parser instead. For now I've just tried to make it handle a few more cases correctly.

For more details of valid and invalid links in markdown, see https://spec.commonmark.org/0.28/#links

dkhamsing commented:
Thx Mat

1 are all tests passing for you ?
2 could you look in on these cases https://github.com/dkhamsing/awesome_bot/blob/master/.travis.yml

MatMoore commented:
Hi @dkhamsing, actually four tests fail in status_spec and timeout_spec, but they also failed on master. It seems like at least one of them is due to a failure to connect to an external server: `"Failed to open TCP connection to www.cmr.osu.edu:80 (getaddrinfo: nodename nor servname provided, or not known)" to include "execution expired"`

I had unreliable internet yesterday so I thought I'd just push it and see if it passes on travis, but I guess not. I'll see if I can fix it by updating the tests.

I did just notice there was a warning about my regex syntax, so I've just corrected that.

MatMoore commented:
Hi @dkhamsing, have you had a chance to think about https://github.com/dkhamsing/awesome_bot/pull/174? If not I'll need to update the URLs used by the `bin/assets` scripts in order to get this one passing.

I'm quite keen to get this PR merged as the problem breaks my build on another project, so let me know if there's anything I can help with.

dkhamsing commented:
I think we’ll have to defer #174 - let’s try to get tests passing here, do let me know of invalid tests as well and we also need to get the rspec ones green, if possible 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

MatMoore commented:
Ok, I've added 3 commits to this branch that get all the tests passing.

These are the problematic URLs I've changed:

|URL|Assumed old behaviour|Current behaviour|Change|
|--|--|--|--|
|http://www.cmr.osu.edu/browse/datasets|Times out|Domain no longer resolves, instantly fails|Replace with http://1.2.3.4|
|http://okeowoaderemi.com/site/assets/files/1103/zf2-flowchart.jpg|Contains an invalid header with unicode quote marks in|No longer returns the header|Remove test|
|https://godoc.org/github.com/ipfs/go-libp2p-crypto|Returns a redirect, with a relative URL in the `Location` header|No longer redirects|Replace with httpbin URL|

dkhamsing commented:
wow thanks!

dkhamsing commented on lib/awesome_bot/links.rb:
```diff
@@ -45,9 +45,10 @@ def links_find(content, url_base=nil)
     def get_relative_links(content, base)
-      links = content.scan /\].*?\)/
+      links = []
+      content.scan(/\[[^\]]+\] \( <? ([^)<>]+) >? \)/x) { |groups| links << groups.first }
```
@MatMoore could u add a comment explains this pls!

dkhamsing commented:
@MatMoore ready to merge after you add just a blurb explaining the new parsing, if appropriate - thx

MatMoore commented:
Cool, I've added a comment in.

dkhamsing commented:
Thx @MatMoore! Released as `awesome_bot 1.18.0`.
